crates/chia-datalayer/src/merkle.rs

```diff
@@ -0,0 +1,1902 @@
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Streamable)]
+pub struct KvId(i64);
+
+#[cfg(feature = "py-bindings")]
+impl IntoPy<PyObject> for KvId {
+    fn into_py(self, py: Python<'_>) -> PyObject {
+        self.0.into_py(py)
+    }
+}
+
+impl std::fmt::Display for KvId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+#[derive(Debug, Error, PartialEq, Eq)]
+pub enum Error {
+    #[error("unknown NodeType value: {0:?}")]
+    UnknownNodeTypeValue(u8),
+
+    #[error("unknown dirty value: {0:?}")]
+    UnknownDirtyValue(u8),
+
+    // TODO: don't use String here
+    #[error("failed loading metadata: {0}")]
+    FailedLoadingMetadata(String),
+
+    // TODO: don't use String here
+    #[error("failed loading node: {0}")]
+    FailedLoadingNode(String),
+
+    #[error("blob length must be a multiple of block count, found extra bytes: {0}")]
+    InvalidBlobLength(usize),
+
+    #[error("key already present")]
+    KeyAlreadyPresent,
+
+    #[error("requested insertion at root but tree not empty")]
+    UnableToInsertAsRootOfNonEmptyTree,
+
+    #[error("unable to find a leaf")]
+    UnableToFindALeaf,
+
+    #[error("unknown key: {0:?}")]
+    UnknownKey(KvId),
+
+    #[error("key not in key to index cache: {0:?}")]
+    IntegrityKeyNotInCache(KvId),
+
+    #[error("zero-length seed bytes not allowed")]
+    ZeroLengthSeedNotAllowed,
+
+    #[error("block index out of range: {0:?}")]
+    BlockIndexOutOfRange(TreeIndex),
+
+    #[error("node not a leaf: {0:?}")]
+    NodeNotALeaf(InternalNode),
+
+    #[error("from streamable: {0:?}")]
+    Streaming(chia_traits::chia_error::Error),
+
+    #[error("index not a child: {0}")]
+    IndexIsNotAChild(TreeIndex),
+}
+
+// assumptions
+// - root is at index 0
+// - any case with no keys will have a zero length blob
+
+// define the serialized block format
+const METADATA_RANGE: Range<usize> = 0..METADATA_SIZE;
+const METADATA_SIZE: usize = 2;
+// TODO: figure out the real max better than trial and error?
+const DATA_SIZE: usize = 53;
+const BLOCK_SIZE: usize = METADATA_SIZE + DATA_SIZE;
+type BlockBytes = [u8; BLOCK_SIZE];
+type MetadataBytes = [u8; METADATA_SIZE];
+type DataBytes = [u8; DATA_SIZE];
+const DATA_RANGE: Range<usize> = METADATA_SIZE..METADATA_SIZE + DATA_SIZE;
+
+fn streamable_from_bytes_ignore_extra_bytes<T>(bytes: &[u8]) -> Result<T, Error>
+where
+    T: Streamable,
+{
+    let mut cursor = std::io::Cursor::new(bytes);
+    // TODO: consider trusted mode?
+    T::parse::<false>(&mut cursor).map_err(Error::Streaming)
+}
+
+#[repr(u8)]
+#[derive(Clone, Copy, Debug, Hash, Eq, PartialEq, Streamable)]
+pub enum NodeType {
+    Internal = 0,
+    Leaf = 1,
+}
+
+impl NodeType {
+    pub fn from_u8(value: u8) -> Result<Self, Error> {
+        streamable_from_bytes_ignore_extra_bytes(&[value])
+    }
+
+    #[allow(clippy::wrong_self_convention, clippy::trivially_copy_pass_by_ref)]
```

Comment:
This lint exists for a reason, taking it by reference isn't idiomatic and uses 8 bytes for a pointer instead of 1 for the enum itself. You should take self by value here.